### PR TITLE
feat: add geocode-earth provider

### LINKED ISSUE #808: Mapzen is shutting down its geocoder
https://mapzen.com/blog/shutdown/

(This is a placeholder issue. I don't know exactly what needs to happen for this repository. Does the Mapzen provider just need to be de-registered? The Mapzen Search is open source so it's possible that someone could stand up their own instance and use the existing Mapzen geocoder-php provider hook.)

### PR DESCRIPTION
**Pull request by atymic**
This adds a new provider for https://geocode.earth/

They don't seem to have any public API docs, nor a free version of the API available without signup.
I've flicked them an email asking for a test key + API docs.

For the moment all the tests are failing due to this, i'll resolve this once I get the test key

I'm also not sure of the exact process for creating a new provider, so let me know what needs to be fixed 

Resolves #808

Thanks :smile:

### CODE REVIEW

File: src/Provider/GeocodeEarth/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Change Log
+
+The change log describes what is "Added", "Removed", "Changed" or "Fixed" between each release.
+
+## 4.0.0
```
Comment (atymic) [inline src/Provider/GeocodeEarth/CHANGELOG.md]: Not sure what this should be, let me know if it should be different.

Comment (jbelien) [reply src/Provider/GeocodeEarth/CHANGELOG.md]: Yes, I know it's a bit confusing 🤣 

It's a new provider, you can start at `1.0.0`.

**Comment by jbelien** [comment]
Thanks a lot for this ! 🎉 

Could you make sure it passes all the tests ?
I just had a quick look and I think there is just the `.cached_response` files missing in your PR.

**Comment by atymic** [comment]
> Could you make sure it passes all the tests ?
> I just had a quick look and I think there is just the `.cached_response` files missing in your PR.

Just waiting for them to drop me some test API keys :)

File: src/Provider/GeocodeEarth/CHANGELOG.md (hunk 1 above)
Comment (atymic) [reply src/Provider/GeocodeEarth/CHANGELOG.md]: Ok, will do tomorrow :)

File: src/Provider/GeocodeEarth/Tests/GeocodeEarthTest.php [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+        $this->assertInstanceOf('Geocoder\Model\AddressCollection', $results);
+        $this->assertCount(5, $results);
+
+        /** @var \Geocoder\Model\Address $result */
+        $result = $results->first();
+        $this->assertInstanceOf('\Geocoder\Model\Address', $result);
+        $this->assertEquals(54.048411999999999, $result->getCoordinates()->getLatitude(), '', 0.001);
+        $this->assertEquals(-2.7989549999999999, $result->getCoordinates()->getLongitude(), '', 0.001);
+        $this->assertEquals(1, $result->getStreetNumber());
```
Comment (atymic) [inline src/Provider/GeocodeEarth/Tests/GeocodeEarthTest.php]: I've updated these tests to match responses from this new provider. Just curious if there are any guidelines for writing tests?

**Comment by atymic** [comment]
The tests seem to be failing on php `< 7.1`, maybe due to serialisation changes?
Not sure how to best resolve this.

**Comment by jbelien** [comment]
To fix the failed test, add `<server name="GEOCODE_EARTH_API_KEY" value="YOUR_GEOCODE_EARTH_API_KEY" />` also in the main [phpunit.xml.dist](https://github.com/geocoder-php/Geocoder/blob/master/phpunit.xml.dist).

**Comment by atymic** [comment]
Done, thanks @jbelien :)

**Comment by atymic** [comment]
Is there anything else required for this PR?

**Comment by jbelien** [comment]
Sorry, didn't have time to properly have a look at it yet!
I'll do that ASAP :)

**Comment by atymic** [comment]
No worries, sorry to bug 😄

File: src/Provider/GeocodeEarth/Tests/GeocodeEarthTest.php (hunk 2 above)
Comment (jbelien) [reply src/Provider/GeocodeEarth/Tests/GeocodeEarthTest.php]: No really, as long as you test "everything" it's a good test :)

File: src/Provider/GeocodeEarth/Tests/IntegrationTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+use Geocoder\Provider\GeocodeEarth\GeocodeEarth;
+use Http\Client\HttpClient;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class IntegrationTest extends ProviderIntegrationTest
+{
+    protected $skippedTests = [
+        'testReverseQueryWithNoResults' => 'We weirdly find stuff here...',
```
Comment (jbelien) [inline src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: Could you explain what you call "weird stuff ?

Comment (atymic) [reply src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: Sorry, this was copy pasted from the old mapzen provider.
There's two results here, though the integration test expects none. This is consistent with the old mapzen results for this test (and why it was skipped there as well)

Comment (jbelien) [reply src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: Could you copy the results here ?
It's just to have an idea of what kind of results this is.

Comment (atymic) [reply src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: There failure is:
```
Failed asserting that 2 matches expected 0.
Expected :0
Actual   :2
```

Results: https://gist.github.com/atymic/781d61b120f481d087064e59ee9e0459

Comment (atymic) [reply src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: image

Comment (jbelien) [reply src/Provider/GeocodeEarth/Tests/IntegrationTest.php]: We usually don't touch/filter what the API returns.
The goal of this project is to provide easy access to geocoding API, not to decide what should or shouldn't be returned. :)

**Comment by jbelien** [comment]
Thanks a lot for the PR @atymic !

@Nyholm We have a new provider 🎉 Could you create the new sub-package ? Thanks.

**Comment by atymic** [comment]
No worries 😄

**Pull request merged**